### PR TITLE
Add network state API

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: shared/api/network.go [hunk 1]
```diff
@@ -58,3 +58,23 @@ type NetworkLease struct {
 	Address  string `json:"address" yaml:"address"`
 	Type     string `json:"type" yaml:"type"`
 }
+
+// NetworkStateStats represents statistics about an interface
+//
+// API extension: network_state
+type NetworkStateStats struct {
```
Comment (stgraber) [inline shared/api/network.go]: This should be renamed to NetworkStateCounters for consistency with containerStateNetworkCounters and its fields should also be renamed to match the container equivalent of this:

```golang
type ContainerStateNetworkCounters struct {
    BytesReceived   int64 `json:"bytes_received" yaml:"bytes_received"`
    BytesSent       int64 `json:"bytes_sent" yaml:"bytes_sent"`
    PacketsReceived int64 `json:"packets_received" yaml:"packets_received"`
    PacketsSent     int64 `json:"packets_sent" yaml:"packets_sent"`
}
```

File: shared/api/network.go [hunk 2]
```diff
@@ -58,3 +58,23 @@ type NetworkLease struct {
...
+	PacketsReceived    int64 `json:"packets_received" yaml:"packets_received"`
+}
+
+// NetworkState represents the state of an interface
+//
+// API extension: network_state
+type NetworkState struct {
+	MTU       int               `json:"mtu" yaml:"mtu"`
+	Hwaddr    string            `json:"hwaddr" yaml:"hwaddr"`
+	Addresses []string          `json:"addresses" yaml:"addresses"`
```
Comment (stgraber) [inline shared/api/network.go]: This should be using a struct a NetworkStateAddress struct that's similar to ContainerStateNetworkAddress:
```golang
type ContainerStateNetworkAddress struct {
    Family  string `json:"family" yaml:"family"`
    Address string `json:"address" yaml:"address"`
    Netmask string `json:"netmask" yaml:"netmask"`
    Scope   string `json:"scope" yaml:"scope"`
}
```

File: shared/api/network.go [hunk 3]
```diff
@@ -58,3 +58,23 @@ type NetworkLease struct {
...
+	BytesReceived      int64 `json:"bytes_received" yaml:"bytes_received"`
+	PacketsTransmitted int64 `json:"packets_transmitted" yaml:"packets_transmitted"`
+	PacketsReceived    int64 `json:"packets_received" yaml:"packets_received"`
+}
+
+// NetworkState represents the state of an interface
+//
+// API extension: network_state
+type NetworkState struct {
+	MTU       int               `json:"mtu" yaml:"mtu"`
```
Comment (stgraber) [inline shared/api/network.go]: This should be called `Mtu` for consistency (an unfortunate one though) with `ContainerStateNetwork`

File: shared/api/network.go [hunk 4]
```diff
@@ -58,3 +58,23 @@ type NetworkLease struct {
...
+	BytesTransmitted   int64 `json:"bytes_transmitted" yaml:"bytes_transmitted"`
+	BytesReceived      int64 `json:"bytes_received" yaml:"bytes_received"`
+	PacketsTransmitted int64 `json:"packets_transmitted" yaml:"packets_transmitted"`
+	PacketsReceived    int64 `json:"packets_received" yaml:"packets_received"`
+}
+
+// NetworkState represents the state of an interface
+//
+// API extension: network_state
+type NetworkState struct {
```
Comment (stgraber) [inline shared/api/network.go]: Fields in the struct should be sorted

File: shared/api/network.go [hunk 5]
```diff
@@ -58,3 +58,23 @@ type NetworkLease struct {
...
+	MTU       int               `json:"mtu" yaml:"mtu"`
+	Hwaddr    string            `json:"hwaddr" yaml:"hwaddr"`
+	Addresses []string          `json:"addresses" yaml:"addresses"`
+	Stats     NetworkStateStats `json:"stats" yaml:"stats"`
+}
```
Comment (stgraber) [inline shared/api/network.go]: For consistency with `ContainerStateNetwork` we probably want to add `State` and `Type` here too.
`Type` will be a bit redundant as NetworkGet effectively gets you that but it doesn't hurt to duplicate it here.

**Comment by monstermunchkin** [comment]
Added the doc part. Also, I removed the (duplicate) NetworkState structs and am now using the existing ones.

**Comment by monstermunchkin** [comment]
I re-introduced the network structs, and removed the `omitempty`.

**Comment by stgraber** [comment]
Looks like you broke `lxc info` :)

I'll take a look and push what should be a pretty easy fix for it.

**Comment by stgraber** [comment]
Added a CLI subcommand for it and added to the Go API too.

**Comment by stgraber** [comment]
@brauner please review since both Thomas and I contributed to that one :)

**Comment by stgraber** [comment]
@brauner

File: lxc/network.go [hunk 6]
```diff
@@ -705,6 +709,73 @@ func (c *cmdNetworkGet) Run(cmd *cobra.Command, args []string) error {
...
+	global  *cmdGlobal
+	network *cmdNetwork
+}
+
+func (c *cmdNetworkInfo) Command() *cobra.Command {
+	cmd := &cobra.Command{}
+	cmd.Use = i18n.G("info [<remote>:]<network>")
+	cmd.Short = i18n.G("Get runtime information of networks")
+	cmd.Long = cli.FormatSection(i18n.G("Description"), i18n.G(
+		`Get runtime information on networks`))
```
Comment (brauner) [inline lxc/network.go]: It writes `on` here but `of` right before. :)

File: lxd/networks.go [hunk 7]
```diff
@@ -806,6 +825,21 @@ func networkShutdown(s *state.State) error {
 	return nil
 }
 
+func networkStateGet(d *Daemon, r *http.Request) Response {
+	name := mux.Vars(r)["name"]
+
+	// Get some information
+	osInfo, _ := net.InterfaceByName(name)
+	_, dbInfo, _ := d.cluster.NetworkGet(name)
```
Comment (brauner) [inline lxd/networks.go]: Ignoring errors is safe here?

Comment (stgraber) [reply lxd/networks.go]: Yeah, that's normal. Networks can be defined on the system (osInfo) or in the database (dbInfo) or both.
Them being missing in one or the other is fine, we only fail if both are missing.

File: lxc/network.go (hunk 6 above)
Comment (stgraber) [reply lxc/network.go]: Oops, fixing

**Comment by stgraber** [comment]
@brauner fixed (or commented)

**Pull request merged**